File: src/prod/src/managed/Api/src/System/Fabric/ImageStore/FileLock.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.ImageStore
{
    using System.Fabric.Common;
    using System.Fabric.Common.Tracing;
    using System.IO;
    using System.Diagnostics;
    using System.Threading;
    using System.Fabric.Strings;

    internal class FileLock : IDisposable
    {
        public const string ReaderLockExtension = ".ReaderLock";
        public const string WriterLockExtension = ".WriterLock";
        private const string TraceTag = "ManagedFileLock";
        // scope has to be defined because Trace is defined for both System.Fabric.Common.Tracing and System.Diagnostics
        private FabricEvents.ExtensionsEvents traceSource = new FabricEvents.ExtensionsEvents(FabricEvents.Tasks.ImageStoreClient);

        private FileStream writerStream;
        private FileStream readerStream;
        private string path;
        private bool isReaderLock;
        private string writerLockPath;
        private string readerLockPath;        
        private string operation;

        public FileLock(string path, bool isReaderLock)
        {
            this.path = path;            
            this.isReaderLock = isReaderLock;
            this.writerLockPath = this.path + FileLock.WriterLockExtension;
            this.readerLockPath = this.path + FileLock.ReaderLockExtension;            
            this.operation = this.isReaderLock ? "reader" : "writer";
        }
        
        public static bool DoesWriterLockExist(string path)
        {
            string writerLockPath = path + FileLock.WriterLockExtension;
            return FabricFile.Exists(writerLockPath);
        }

        public bool Acquire()
        {
            Exception exception = null;
            FileStream localReaderStream = null;
            FileStream localWriterStream = null;
            try
            {
                if (this.isReaderLock)
                {
#if DotNetCoreClrLinux
                    this.CreateDirectoryPath(this.readerLockPath);
                    localReaderStream = File.Open(this.readerLockPath, FileMode.OpenOrCreate, FileAccess.Read, FileShare.Read);
#else
                    localReaderStream = FabricFile.Open(this.readerLockPath, FileMode.OpenOrCreate, FileAccess.Read, FileShare.Read);
#endif
                    if (FabricFile.Exists(this.writerLockPath))
                    {
                        this.traceSource.WriteWarning(
                            TraceTag,
                            "Failed to acquire reader lock for {0} since writer lock is found. This could indicate that the file is being written or corrupted.",
                            this.path);

                        throw new FabricImageStoreException(
                            StringResources.Error_CorruptedImageStoreObjectFound,
                            FabricErrorCode.CorruptedImageStoreObjectFound);
                    }
                }
                else
                {
#if DotNetCoreClrLinux
                    this.CreateDirectoryPath(this.readerLockPath);
                    localReaderStream = File.Open(this.readerLockPath, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None);
                    this.CreateDirectoryPath(this.writerLockPath);
                    localWriterStream = File.Open(this.writerLockPath, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None);
#else
                    localReaderStream = FabricFile.Open(this.readerLockPath, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None);
                    localWriterStream = FabricFile.Open(this.writerLockPath, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None);
#endif
                }
            }
            catch (Exception e)
            {
                exception = e;
            }

            if (exception == null)
            {
                this.readerStream = localReaderStream;
                this.writerStream = localWriterStream;

                this.traceSource.WriteInfo(TraceTag, "Obtained {0} lock for {1}", this.operation, this.path);
                return true;
            }
            else
            {
                if (localReaderStream != null)
                {
                    localReaderStream.Dispose();
                }

                if (localWriterStream != null)
                {
                    localWriterStream.Dispose();
                }

                this.traceSource.WriteWarning(
                    TraceTag,
                    "Failed to acquire {0} lock for {1}. Error = {2}",
                    this.operation,
                    this.path,
                    exception);

                return false;
            }
        }

        public bool Acquire(TimeSpan timeout)
        {
            Stopwatch stopWatch = new Stopwatch();
            bool success;
            
            do
            {
                success = Acquire();
                if (!success && stopWatch.Elapsed < timeout)
                {
                    Thread.Sleep(100);
                }
            } 
            while (!success && stopWatch.Elapsed < timeout);

            return success;
        }

        public bool Release()
        {            
            if (this.readerStream != null)
            {             
                try
                {
                    this.readerStream.Dispose();
                    this.readerStream = null;
                    FabricFile.Delete(this.readerLockPath, deleteReadonly: true);
                }
                catch (Exception e)
                {
                    this.traceSource.WriteInfo(TraceTag, "Unable to delete {0} because of {1}", this.readerLockPath, e);
                }
            }

            if (this.writerStream != null)
            {
                this.writerStream.Dispose();
                this.writerStream = null;
                FabricFile.Delete(this.writerLockPath, deleteReadonly: true);
            }

            this.traceSource.WriteInfo(TraceTag, "Released {0} lock on {1}", this.operation, this.path);

            return true;
        }

        public void Dispose()
        {
            this.Release();
            GC.SuppressFinalize(this);
        }

#if DotNetCoreClrLinux
        private void CreateDirectoryPath(string path)
        {
            string directoryName = FabricPath.GetDirectoryName(path);
            if ((directoryName != null) && (directoryName.Length > 0) && (!FabricDirectory.Exists(directoryName)))
            {
                FabricDirectory.CreateDirectory(directoryName);
            }
        }
#endif
    }
}

File: src/prod/src/Common/FileLock.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

namespace Common
{
    using namespace std;

    static Common::StringLiteral const FileLockSource = "FileLock";

    template <bool IsReaderLock>
    const std::wstring FileLock<IsReaderLock>::ReaderLockExtension = L".ReaderLock";
    template <bool IsReaderLock>
    const std::wstring FileLock<IsReaderLock>::WriterLockExtension = L".WriterLock";

    template <bool IsReaderLock>
    FileLock<IsReaderLock>::FileLock(std::wstring const & path)
        : readFile_(), 
          writeFile_(),
          path_(path),
          operation_(IsReaderLock ? L"reader" : L"writer")
    {
    }

    template <bool IsReaderLock>
    FileLock<IsReaderLock>::~FileLock()
    {
        Release();
    }

    template <bool IsReaderLock>
    std::wstring FileLock<IsReaderLock>::GetReaderLockPath(std::wstring const & path)
    {
        return path + FileLock::ReaderLockExtension;
    }

    template <bool IsReaderLock>
    std::wstring FileLock<IsReaderLock>::GetWriterLockPath(std::wstring const & path)
    {
        return path + FileLock::WriterLockExtension;
    }
    
    template <bool IsReaderLock>
    ErrorCode FileLock<IsReaderLock>::AcquireImpl()
    {   
        unique_ptr<File> localReadFile = make_unique<File>();
        unique_ptr<File> localWriteFile = make_unique<File>();

        wstring writerLockPath = FileLock::GetWriterLockPath(this->path_);
        wstring readerLockPath = FileLock::GetReaderLockPath(this->path_);

        ErrorCode errorCode = ErrorCode(ErrorCodeValue::Success);

        bool isReaderLock = IsReaderLock;
        
        if (isReaderLock)
        {
            errorCode = localReadFile->TryOpen(
                readerLockPath,
                FileMode::OpenOrCreate,
                FileAccess::Read,
                FileShare::Read,
                FileAttributes::None);
            if (!errorCode.IsSuccess())
            {
                Trace.WriteWarning(FileLockSource, "Failed to acquire ReaderLock for path {0}. LockType = {1} Error = {2}", path_, operation_, errorCode);

                if (errorCode.IsWin32Error(ERROR_SHARING_VIOLATION))
                {
                    return ErrorCodeValue::SharingAccessLockViolation;
                }

                return errorCode;
            }
            
            if (File::Exists(writerLockPath))
            {
                Trace.WriteWarning(FileLockSource, "CorruptedImageStoreObjectFound found for path {0}. LockType = {1}", path_, operation_);
                return ErrorCodeValue::CorruptedImageStoreObjectFound;
            }

            readFile_ = move(localReadFile);

        }
        else
        {
            errorCode = localReadFile->TryOpen(
                readerLockPath,
                FileMode::OpenOrCreate,
                FileAccess::Write,
                FileShare::None,
                FileAttributes::None);
            if (!errorCode.IsSuccess())
            {
                Trace.WriteInfo(FileLockSource, "Failed to acquire ReaderLock for path {0}. LockType = {1} Error = {2}", path_, operation_, errorCode);
                
                if (errorCode.IsWin32Error(ERROR_SHARING_VIOLATION))
                {
                    return ErrorCodeValue::SharingAccessLockViolation;
                }

                return errorCode;
            }

            errorCode = localWriteFile->TryOpen(
                writerLockPath,
                FileMode::OpenOrCreate,
                FileAccess::Write,
                FileShare::None,
                FileAttributes::None);
            if (!errorCode.IsSuccess())
            {
                Trace.WriteInfo(FileLockSource, "Failed to acquire WriterLock for path {0}. LockType = {1} Error = {2}", path_, operation_, errorCode);

                if (errorCode.IsWin32Error(ERROR_SHARING_VIOLATION))
                {
                    errorCode = ErrorCodeValue::SharingAccessLockViolation;
                }

                return errorCode;
            }

            readFile_ = move(localReadFile);
            writeFile_ = move(localWriteFile);
        }

        Trace.WriteInfo(FileLockSource, "Obtained {0} lock for {1}", operation_, path_);            

        return errorCode;
    }

    template <bool IsReaderLock>
    bool FileLock<IsReaderLock>::Release()
    {
        if(writeFile_)
        {
            writeFile_.reset();
            wstring writerLockPath = FileLock::GetWriterLockPath(this->path_);
            ErrorCode error = File::Delete2(writerLockPath);
            Trace.WriteTrace(
                error.ToLogLevel(),
                FileLockSource,
                "Delete lock file {0}. Error:{1}", writerLockPath, error);

            if(!error.IsSuccess())
            {
                return false;
            }
        }

        if(readFile_)
        {
            readFile_.reset();
            wstring readerLockPath = FileLock::GetReaderLockPath(this->path_);
            ErrorCode error = File::Delete2(readerLockPath);            
            Trace.WriteNoise(
                FileLockSource, 
                "Delete lock file {0}. Error:{1}", readerLockPath, error);
            
        }
       
        Trace.WriteInfo(FileLockSource, "Released {0} lock on {1}", operation_, path_);

        return true;
    }

    template <bool IsReaderLock>
    ErrorCode FileLock<IsReaderLock>::Acquire(TimeSpan timeout)
    {
        TimeoutHelper timeoutHelper(timeout);
        ErrorCode lastError;

        do
        {
            lastError = AcquireImpl();

            if (!timeoutHelper.IsExpired && !lastError.IsSuccess())
            {
                Sleep(100);
            }
        } while (!timeoutHelper.IsExpired && !lastError.IsSuccess());

        return lastError;
    }

    template class FileLock<true>;
    template class FileLock<false>;
}
